Isse/7900 timezone selection - Rewrite Timezone selection using Kotlin with BottomSheetDialogFragment

ravishanker opened the pull request:
Timezone selection from site settings in Android app is outdated in comparison with WordPress Core and iOS.  This PR brings android on par with them by adding missing features like grouping by continents and manual offsets.  Please see #7900 for more details on this issue.

It also improves the user experience by introducing bottom sheet,  better search and timezone information similar to android system timezone selection feature.  As below.

It is also a rewrite using modern android practices like Kotlin and MVVM pattern from current everything in one Java file implementation.

Fixes #7900 

To do:
- <strike>Refine timezone offset information</strike>
- <strike>Refine search feature</strike>
- <strike>Suggested timezone</strike>

To test:
Navigate to My Site -> Settings -> Timezone

<table><tr><td><img width="320" src="https://user-images.githubusercontent.com/990349/111855129-216f7c00-8977-11eb-9861-f3ccf7ede1ee.png"></td>
<td><img width="320" src="https://user-images.githubusercontent.com/990349/111855148-3fd57780-8977-11eb-9b9b-b17a8bf32596.png"></td>
<td><img width="320" src="https://user-images.githubusercontent.com/990349/111119414-38e8e680-85be-11eb-999b-01162e62dcb4.png"></td></tr></table>


PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
New Timezone Selection Bottom Sheet

Refactor SiteSettingsTimzoneDialog to mvvm pattern by
add new modal BottomSheetDialogFragment
add view model
create layouts for fragment, list and item
create Adapter
create ViewHolder
update dagger component and module

ravishanker committed:
Update SiteSettingsFragment.java

hook new TimezoneBottomSheet with SiteSettingsFragment

ravishanker committed:
Group timezones by continent

Group timezones by continent
Add new header view holder
parse json with Gson
update Adapter

ravishanker committed:
Replace SearchView with material outlined text field

update SearchView with material outlined text field
bump materialVersion to 1.30
bump coreVersion to 1.3.2
add core-ktx
rename isRtl private variable to isRightToLeft to fix conflict in WPEditTextWithChipsOutlined

ravishanker committed:
Two line summary Timezone list

Removed the api call
Use android TimeZone class to find available zones
Clean up

ravishanker committed:
Remove api related files

delete Timezones data classes

ravishanker committed:
Display timezone variation for android versions

Display two line list item timezone with offset on android version code greater than O(26)
Display timezone only for versions below O
Improved search handling

ravishanker committed:
Timezones on all android versions

Using java.time package to get timezone information
Enabled core library desugaring to utilize jav.time package
Improved logic and display

khaykov commented:
Hey @ravishanker 👋 Thanks for the PR! I checked the functionality of timezone picker looks like it works pretty well. 

Pulling timezones from the `TimeZone` class is pretty great, since we don't have to deal with any networking, but I have one concern about this approach - `TimeZone` class has a lot of timezones, and a there is a good number of zones that are not relevant (if you scroll through the list there is a lot of weird stuff). What do you think about pulling the timezones from our endpoint and using `TimeZone` to give them more information (what you are currently using in the list)?

ravishanker commented:
> Hey @ravishanker 👋 Thanks for the PR! I checked the functionality of timezone picker looks like it works pretty well.
> 
> Pulling timezones from the `TimeZone` class is pretty great, since we don't have to deal with any networking, but I have one concern about this approach - `TimeZone` class has a lot of timezones, and a there is a good number of zones that are not relevant (if you scroll through the list there is a lot of weird stuff). What do you think about pulling the timezones from our endpoint and using `TimeZone` to give them more information (what you are currently using in the list)?

Hi @khaykov Thanks for the feedback 🙏.  I'm looking at filtering entries to remove all those extras  For e.g. limiting to Los Angeles -> Pacific Time (GMT-8:00) or Sydney -> Australian Eastern Time (GMT+11).  Perhaps, making an API call could be an option if we can't figure a way to filter these entries.

ravishanker committed:
Filter timezones

filter timezones to limit the data to continents only

ravishanker committed:
Layout update

adjust padding on header and list items

ravishanker committed:
offset display

update GMT offset display

ravishanker commented:
@khakov 🙋‍♂️ Finally, managed to filter the data, update layout without the need to fallback on API.  Please take a look again. Thanks

khaykov commented:
@ravishanker Thanks for the changes, weird timezones are gone 👍 Thinking about functionality and use cases, what do you think about including timezone name (the second row in the list elements) into the search? I see cases where folks would know their timezone name, but not necessarily what major city it might be associated with. Eg. living in the middle of US, you might search for "Central Time" instead of a "Chicago" (or some other major city). In addition, we still probably would want to include manual (and searchable) UTC offsets for convenience.

ravishanker commented:
@khaykov 👋 Thanks for the review 🙏.  That's a good idea to include second row into search field.  I'll add back manual UTC offsets.  
- Should continents be included as well into searchable?
- We're showing GMT prefix (GMT+6:00) instead of UTC, should that be GMT or UTC?

ravishanker committed:
Timezone suggestion

add time zone suggestion
include time zone offset in search results
add manual offsets

ravishanker commented:
@khaykov I've added the following now

- timezone name in search results
- suggested time zone under the search box (see updated screenshot above)
However, the manual offsets are not working.  Looks like the api used in android is not handling manual offsets.  It is not accepting and updating the manual offsets.  I've even tried adding the list from api, but to no avail unless I'm missing something here!

ravishanker committed:
Revert to using wp timezone api

- replace TimeZone use with wp timezone api
- add manual offsets from the api
- fix overlapping offset on time in second row
- set second row to gone for manual offsets items

ravishanker commented:
@khaykov  :wave: What I found on investigating the issue is
when you select a manual offset for e.g UTC+7, the back-end is updating it is as "gmt_offset: 7" and
when site settings is fetched it is actually returning empty ""  

see WpComSiteSettings.java, pushWpSettings method line 382 and fetchWpSettings method on line 168)
It is working alright when selecting anything other than manual offsets though.
I've also now reverted from use of TimeZone class to calling AP and augmenting it with timezone offset and time.

ravishanker committed:
handle gmt_offset to UTC

Android returns manual offsets as gmt_offset and not as UTC when manual offsets are selected in the form UTC+/- form

adds a work around for the Timezone display by mapping gmt_offset to UTC

ravishanker committed:
Remove previous implementation

- Remove previous SiteSettingsTimezoneDialog
- Clean up references in SiteSettingsFragment

ravishanker committed:
Improved search experience

- add ime action and keycode handling

ravishanker committed:
Merge branch 'develop' into isse/7900-timezone-selection

ravishanker committed:
format zone display

- Removed "/" and "_" from timezone display

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneBottomSheet.kt:
```diff
@@ -0,0 +1,206 @@
+        fun newInstance(timezone: String): SiteSettingsTimezoneBottomSheet =
+                SiteSettingsTimezoneBottomSheet().apply {
+                    arguments = Bundle().apply {
+                        putString(KEY_TIMEZONE, timezone)
```
Looks like `KEY_TIMEZONE` migrated from previous implementation, but I don't think we are using it anywhere.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneBottomSheet.kt:
```diff
@@ -0,0 +1,206 @@
+        setSelectedTimezone(timezone)
+    }
+
+    private var searchJob: Job? = null
```
This is a pretty minor thing, but what do you think about moving this `Job` into ViewModel? So instead of calling `search()` or `updateTimezonesFromInput()` when text changes, we would just let the VM know that the text in search filed have changed, and it can do its magic.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneBottomSheet.kt:
```diff
@@ -0,0 +1,206 @@
+            binding?.btnTimezoneSuggestion?.text = it
+        })
+
+        timezoneViewModel.timezoneSearch.observe(viewLifecycleOwner, {
```
From the perspective of the list, `timezones` and `timezoneSearch` are the same, do you think we can only observe `timezones` and update VM logic to modify `timezones` on search?

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneBottomSheet.kt:
```diff
@@ -0,0 +1,206 @@
+        }
+    }
+
+    private fun showEmptyView(show: Boolean) {
```
Pretty minor comment, but to make things a bit more clear, what do think about renaming `showEmptyView` to something like `toggleEmptyViewVisibility`. To me personally, `showEmptyView` implies exactly that - to show empty view, so having `show` attribute is a bit weird. Same goes for `showProgressView` method.

khaykov replied:
Pretty minor comment, but to make things a bit more clear, what do think about renaming `showEmptyView` to something like `toggleEmptyViewVisibility`. To me personally, `showEmptyView` implies exactly that - to show empty view, so having `show` attribute is a bit weird. Same goes for `showProgressView` method.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneBottomSheet.kt:
```diff
@@ -0,0 +1,206 @@
+            }
+
+            btnTimezoneSuggestion.setOnClickListener { it as MaterialButton
+                callback?.onSelectTimezone(it.text.toString())
```
I see we have a similar code in `setSelectedTimezone`, I wonder if we can refactor it a bit by wiring it through ViewModel. Maybe observe something like `selectedTimezone` from VM, and notify it that timezone was selected (either from suggestion or adapter).

We can go even further, and separate call to callback and `dismiss()` (by observing something like `dismissBottomSheet`) and reacting to it. 

This might be going a bit too granular, but we can benefit from separating this logic when testing VM. We would be able to confirm that Timezone WAS selected, and dialog WAS dismissed. With current implementation we can't test what would happen after you click on the timezone.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModel.kt:
```diff
@@ -0,0 +1,211 @@
+class SiteSettingsTimezoneViewModel @Inject constructor() : ViewModel() {
+    private val timezonesList = mutableListOf<TimezonesList>()
+
+    private val _showEmpty = MutableLiveData<Boolean>()
```
Very minor thing - historically, we call mutable and regular live data the same way + underscore (eg. `_showEmpty` will be `_ showEmptyView` :)

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModel.kt:
```diff
@@ -0,0 +1,211 @@
+        _timezones.postValue(timezonesList)
+    }
+
+    fun getTimezones(context: Activity) {
```
Dow we actually need an `Activity` here? Or the `context` will be enough?

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModel.kt:
```diff
@@ -0,0 +1,211 @@
+
+
+    fun searchTimezones(query: CharSequence?) {
+        searchInput.value = query.toString()
```
`query` is marked as nullable, so if at any point it will actually be a null, we will filter with a "null" string, which might not be what want :)

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModel.kt:
```diff
@@ -0,0 +1,211 @@
+        }
+    }
+
+    enum class Continents(val s: String) {
```
For readability, we try to avoid ambitious attributes names like `s`. Also, it looks like the continent names from here will be visible in UI, so they need to be in strings resources.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModel.kt:
```diff
@@ -0,0 +1,211 @@
+                addTimezoneItems(jsonTimezonesByContinent.getJSONArray(it.s))
+            }
+
+            timezonesList.add(TimezoneHeader("Manual Offsets"))
```
Will this string be visible? If so, we would probably like to add it to `strings.xml`

khaykov commented on WordPress/src/main/res/drawable/bg_rectangle_grey_stroke_radius_4dp.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Not really sure where this file is used?

khaykov commented on WordPress/src/main/res/layout/site_settings_timezone_bottom_sheet_list_header.xml:
```diff
@@ -0,0 +1,13 @@
+    android:textAppearance="?attr/textAppearanceHeadline6"
+    tools:text="Africa" />
+
+
```
A bit of extra lines here :)

khaykov commented on WordPress/src/main/res/layout/site_settings_timezone_bottom_sheet_list_item.xml:
```diff
@@ -0,0 +1,49 @@
+        android:layout_height="wrap_content"
+        android:textAppearance="?attr/textAppearanceBody2"
+        android:maxLines="1"
+        android:textColor="@color/zs_request_list_light_text_color"
```
We probably would want to use something like `?attr/wpColorOnSurfaceMedium` here, so it will be readable in dark mode. I think `zs_request_list_light_text_color` comes from one of the 3rd party libraries.

khaykov reviewed (commented):
Thank you for improving the TimeZone picker, @ravishanker ! I finished the first pass, checked the code and functionality, and I didn't found any major issues. I found couple of minor issues and have some suggestions. Let me know what you think!

Visual issues:

Bottom of the list is not accesible when keyboard is visible:
[![Image from Gyazo](https://i.gyazo.com/9f0257aa268a276d54d02bfee8639d67.png)](https://gyazo.com/9f0257aa268a276d54d02bfee8639d67)

The ripple animation is affected by side padding, and does not extend to full row:
[![Image from Gyazo](https://i.gyazo.com/0e80d30357fd05b7f44137c5b1615c04.gif)](https://gyazo.com/0e80d30357fd05b7f44137c5b1615c04)

When the list expanded there is a weird gap under it:
[![Image from Gyazo](https://i.gyazo.com/12996b664417b29284ac52c5cab23b70.png)](https://gyazo.com/12996b664417b29284ac52c5cab23b70)

In landscape mode the bottom sheep is collapsed. This might be a default behavior, but we probably want it to be expanded.
[![Image from Gyazo](https://i.gyazo.com/31910dd82285aac029d1d500572b753c.png)](https://gyazo.com/31910dd82285aac029d1d500572b753c)

Suggested timezone looks a bit weird in the center, what do you think about moving it next to the label or sticking it to the end?
[![Image from Gyazo](https://i.gyazo.com/f2f73367c77eecd790472fbd7a0f7384.png)](https://gyazo.com/f2f73367c77eecd790472fbd7a0f7384)

The scrollbar is not visible.

JavonDavis commented:
Hey @ravishanker! Bumping this to `17.1`, if you'd like to get it in for this release please merge this into release17.0 and then DM me to release a new Beta version.

ravishanker committed:
UI updates to address review comments

- fix keyboard covering search results by dismissing it
- fix ripple effect not extending full width
- Remove bottom margin on list expanded
- expand bottom sheet on landscape mode
- Align suggested timezone  to start with the Suggestion label
- show scrollbar

ravishanker committed:
Code updates to address review comments

- Remove unused KEY_TIMEZONE constant
- Refactor search input
- Remove timezoneSearch and use timezones for search
- Rename variable and function names
- Refactor selectedTimezone and dismiss with view model
- Replace strings with string resources
- delete unused xml drawable
- remove extra lines in xml
- replace textColor

ravishanker committed:
strings array

set translatable=false on strings

ravishanker commented:
@khaykov  I've addressed all the PR review comments from above.  Please take a look again.  Thanks.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModel.kt:
```diff
@@ -28,34 +33,55 @@ import java.time.zone.ZoneRulesException
 
+    fun onTimezoneSelected(timezone: String) {
+        _selectedTimezone.value = timezone
+        _dismissBottomSheet.value = Unit
```
TIP: There is a convenient `call()` or `asyncCall()` methods for situations like this :)

ravishanker replied:
Thanks for the tip, will use it

khaykov commented:
Thank you for the changes, @ravishanker! Looks better now 👍 
I still see couple of visual issues:
- Keyboard blocks the list (scroll down to the bottom of the list and open the keyboard).
- Dialog is opened in collapsed state in landscape orientation.

Also there is a certain issue with the way you set up a `TimezoneSelectionCallback` in both current and previous implementation. Could you please explain the problem with the current approach and address it if you can?

ravishanker commented:
> Thank you for the changes, @ravishanker! Looks better now 👍
> I still see couple of visual issues:
> 
>     * Keyboard blocks the list (scroll down to the bottom of the list and open the keyboard).
> 
>     * Dialog is opened in collapsed state in landscape orientation.
> 
> 
> Also there is a certain issue with the way you set up a `TimezoneSelectionCallback` in both current and previous implementation. Could you please explain the problem with the current approach and address it if you can?

- I've implemented keyboard to dismiss when scroll starts.  Do I need to implement softInputMode adjustResize or adjustPan
- I put in a fix for opening bottomsheet expanded in landscape mode.  Seems to work on my device, will check again.
- I'm not sure about the certain issue.  Are you referring to old implementation in Java using targetFragment, onActivityResult and this implementation or the previous version and current version of the present implementation?

ravishanker committed:
Updates for review

- use call() with SingleLiveEvent
- open bottomsheet  expanded when starting out in landscape

ravishanker committed:
add unit test

Added ViewModel test

ravishanker committed:
Refactor callback

- Refactor timezone bottomsheet callback for config changes

ravishanker committed:
timezone callback

- refactor callback setup

ravishanker committed:
Fix viewmodel test

- replace getOrAwaitValue with observeForever
- remove extension function and cleanup imports

ravishanker committed:
Remove desugaring

- disable core library de-sugaring
- refactor to use timezone display

ravishanker committed:
remove strings

- remove redundant strings

khaykov commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/timezone/SiteSettingsTimezoneViewModelTest.kt:
```diff
@@ -0,0 +1,56 @@
+
+    @Test
+    fun testSearchTimezones() {
+        viewModel.timezones.observeForever {}
```
Just for the sanity check and to test the behavior of observer exactly, would you mind assigning result to an object, and asserting it, instead of checking `.value` ?

ravishanker replied:
Updated to assign result to an object and assert it

khaykov reviewed (commented):
Thanks for all great work, Ravi! I checked what's new in the dependency versions you bumped, and looks like there is nothing breaking, and just general improvements, so we are good to go 🚀 

I left one minor comment about the test, and as a last step would you mind pulling the latest changes from `develop` and add this feature to `RELEASE_NOTES.txt`?

ravishanker committed:
fix test

- assign result to an object and assert

ravishanker committed:
Update RELEASE-NOTES.txt

added entry for 17.1 in RELEASE-NOTES

ravishanker committed:
merge upstream

ravishanker committed:
Revert "merge upstream"

This reverts commit 6cc43f1e2a570c8936e7eee1b4d65fdb05e2a5eb.

ravishanker commented:
@khaykov 👋  Updated the RELEASE_NOTES.txt with 17.1 entry

khaykov commented:
@ravishanker Looks like the stories submodule changes snuck in with your last commit. Would you mind reverting it? Thank you!

ravishanker committed:
Merge branch 'develop' into isse/7900-timezone-selection

khaykov commented:
@ravishanker I ran a CI checks, and looks like there are some lint errors. You can check what is wrong by running `./gradlew ktlint`

ravishanker committed:
lint check

fixed lint issues

ravishanker committed:
Merge branch 'develop' into isse/7900-timezone-selection

ravishanker committed:
Merge branch 'develop' into isse/7900-timezone-selection

ravishanker committed:
Update RELEASE-NOTES.txt

add 17.2 milestone entry in release-notes

ravishanker committed:
Fix detekt issues

- fix for TooManyFunctions detekt warning

khaykov committed:
Update RELEASE-NOTES.txt

Added PR link to release notes.

khaykov reviewed (approved):
Thanks for all the changes, Ravi! Looks good to me 🚀
